container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java

```diff
@@ -205,4 +354,402 @@ public static String asHexString(GlobalId gid) {
+                    if (hit.getField(fieldName) == null)
+                        hit.setField(fieldName, field.convert(f));
+                }
+            }
+        }
+
+        private Iterator<Map.Entry<String, Object>> fieldIterator() {
+            return new SummaryFieldDataIterator(hit, type, data.fields().iterator(), index);
+        }
+
+        private Iterator<String> fieldNameIterator() {
+            return new SummaryFieldNameDataIterator(hit, data.fields().iterator(), index);
+        }
+
+        /** A wrapper of a field iterator which skips removed fields. Read only. */
+        private static abstract class SummaryDataIterator<VALUE> implements Iterator<VALUE> {
+
+            private final FastHit hit;
+            private final Iterator<Map.Entry<String, Inspector>> fieldIterator;
+            private final int index;
+
+            /** The next value or null if none, eagerly read because we need to skip removed and overwritten values */
+            private VALUE next;
+
+            SummaryDataIterator(FastHit hit,
+                                Iterator<Map.Entry<String, Inspector>> fieldIterator,
+                                int index) {
+                this.hit = hit;
+                this.fieldIterator = fieldIterator;
+                this.index = index;
+            }
+
+            @Override
+            public boolean hasNext() {
+                return next != null;
+            }
+
+            @Override
+            public VALUE next() {
+                if (next == null) throw new NoSuchElementException();
+
+                VALUE returnValue = next;
+                advanceNext();
+                return returnValue;
+            }
+
+            protected abstract VALUE toValue(Map.Entry<String, Inspector> field);
+
+            protected void advanceNext() {
+                while (fieldIterator.hasNext()) {
+                    Map.Entry<String, Inspector> nextEntry = fieldIterator.next();
+                    String fieldName = nextEntry.getKey();
+                    next = toValue(nextEntry);
+                    if ( next != null &&
+                         ! hit.hasField(fieldName) &&
+                         ! isRemoved(fieldName) &&
+                         ! isAlreadyReturned(fieldName))
+                        return;
+                }
+                next = null;
+            }
+
+            private boolean isRemoved(String fieldName) {
+                return hit.removedFields != null && hit.removedFields.contains(fieldName);
+            }
+
+            private boolean isAlreadyReturned(String fieldName) {
+                for (int i = 0; i < index; i++) {
+                    if (hit.summaries.get(i).type.fieldNames().contains(fieldName))
+                        return true;
+                }
+                return false;
+            }
+
+        }
+
+        private static class SummaryFieldDataIterator extends SummaryDataIterator<Map.Entry<String, Object>> {
```

Comment:
A short comment on the purpose of the class would be good, maybe something like:
Iterate over the fields of a single summary data object.
Also consider changing name to 'SummaryDataFieldIterator'.